parser/parser.go

```diff
@@ -1,15 +1,215 @@
 package parser
 
 import (
+	"fmt"
+	"strconv"
+
 	"github.com/NeowayLabs/abad/ast"
 	"github.com/NeowayLabs/abad/internal/utf16"
+	"github.com/NeowayLabs/abad/lexer"
+	"github.com/NeowayLabs/abad/token"
 )
 
+type (
+	Parser struct {
+		tokens    <-chan lexer.Tokval
+		lookahead []lexer.Tokval
+
+		filename string
+	}
+
+	parserfn func(*Parser) (ast.Node, error)
+)
+
+// used when the tokens is over
+var tokEOF = lexer.Tokval{
+	Type: token.EOF,
+}
+
+var (
+	keywordParsers = map[token.Type]parserfn{}
+	literalParsers = map[token.Type]parserfn{
+		token.Decimal:     parseDecimal,
+		token.Hexadecimal: parseHex,
+	}
+	unaryParsers map[token.Type]parserfn
+)
+
+func init() {
+	unaryParsers = map[token.Type]parserfn{
```

Comment:
Why this one is on the init() and literalParsers is outside it ?

Reply:
I said to you in the whatsapp =)
It's inside init because otherwise there's a declaration loop.

Reply:
The function parseUnary has a reference to `parse` method that needs parseUnary and so on.